src/mongo/shell.ts

```diff
@@ -5,120 +5,201 @@
-					}
+					reject(`Timed out executing MongoDB command "${script}". To use a longer timeout, modify the VS Code 'mongo.shell.timeout' setting.`);
 				},
 				timeout);
-			const disposable = this.onResult.event(result => {
-				disposable.dispose();
 
-				if (result && result.code) {
-					if (result.code === 'ENOENT') {
-						result.message = `This functionality requires the Mongo DB shell, but we could not find it. Please make sure it is on your path or you have set the '${ext.settingsKeys.mongoShellPath}' VS Code setting to point to the Mongo shell executable file (not folder). Attempted command: "${this.execPath}"`;
-					}
-
-					e(result);
-				} else {
-					let lines = (<string>result.result).split(os.EOL).filter(line => !!line && line !== 'Type "it" for more');
-					lines = lines[lines.length - 1] === 'Type "it" for more' ? lines.splice(lines.length - 1, 1) : lines;
-					executed = true;
-					c(lines.join(os.EOL));
-				}
+			// Handle result or error from the console (via fireOnResultOrError)
+			const disposable = this._onResult.event((result: CommandResult) => {
+				clearTimeout(timeoutHandler);
 
-				if (handler) {
-					clearTimeout(handler);
-				}
+				// Only the first result or error will be processed, all others ignored
+				disposable.dispose();
+
+				// Give STDOUT/STDERR a chance to empty
+				setTimeout(
+					() => {
+						if (result.result !== undefined && !this._stderrData) {
+							if (this._mongoShell.killed || this._exited) {
+								throw new Error(prematureExitMessage);
+							}
+
+							let lines = result.result.split(os.EOL).filter(line => !!line);
+							if (lines[lines.length - 1] === mongoShowMoreMessage) {
+								// CONSIDER: Ideally we would ask or allow the user to ask for more data
+								lines = lines.splice(0, lines.length - 1);
+								lines.push("(More)");
+							}
+
+							let text = lines.join(os.EOL);
+							resolve(text);
+						} else {
+							// An error occurred
+							let errorDetails = result.errorMessage || "";
+							if (errorDetails.includes('ENOENT')) {
+								reject(`This functionality requires the Mongo DB shell, but we could not find it. Please make sure it is on your path or you have set the '${ext.settingsKeys.mongoShellPath}' VS Code setting to point to the Mongo shell executable folder and file path.`
+									+ ` ${shellDetails}`
+								);
+								return;
+							}
+
+							if (!errorDetails) {
+								if (this._stderrData) {
+									errorDetails = this._stderrData;
+									this._stderrData = "";
+								} else {
+									errorDetails = "Unknown error";
+								}
+							}
+
+							ext.outputChannel.appendLine(`Shell execution details: ${shellDetails}`);
+
+							// Add some STDOUT/STDERR context if available
+							const maxOutputLengthInMessage = 500;
+							if (this._stderrData) {
+								errorDetails += `${os.EOL}${os.EOL}${this._stderrData}`.slice(0, maxOutputLengthInMessage);
+							}
+							if (this._stdoutData) {
```

Comment:
I would actually print everything to an outputChannel as it happens (mentioned above) and just point them to the outputChannel in this error